File: doc/example_2.f
C Example 2:
C This example differs from example_1 in that we add explicit lights
C and camera.

C Subroutine to ease checking of return codes
      subroutine errchk( ival )
      if (ival.ne.1) write(6,*) 'ERROR!'
      return
      end

      program ex2
      integer pintrn, popen, psphr, pclose, psnap, pshtdn, pamblt,
     $     plight, pcamra

C Light specification information
      real ltloc(3), ltclr(4), ambclr(4)
      data ltloc/0.0, 20.0, 20.0/
      data ltclr/0.7, 0.7, 0.7, 1.0/
      data ambclr/0.3, 0.3, 0.3, 1.0/

C Camera specification information
      real lookfm(3), lookat(3), up(3)
      data lookfm/0.0, 0.0, 10.0/
      data lookat/0.0, 0.0, 0.0/
      data up/0.0, 1.0, 0.0/
      real fovea, hither, yon
      data fovea, hither, yon/ 53.0, -5.0, -15.0 /

      call errchk( pintrn('myrenderer','p3d','example_2.p3d',
     $     'This string will be ignored') )

      call errchk( popen('mygob') )
      call errchk( psphr() )
      call errchk( pclose() )

      call errchk( popen('mylights') )
      call errchk( pamblt( 0, ambclr ) )
      call errchk( plight( ltloc, 0, ltclr ) )
      call errchk( pclose() )

      call errchk( pcamra('mycamera', lookfm, lookat, up, fovea,
     $     hither, yon ) )

      call errchk( psnap('mygob','mylights','mycamera') )

      call errchk( pshtdn() )

      stop
      end

File: doc/example_2.c
/* Example 2:
 * This example differs from example_1 in that we add explicit lights
 * and camera.
 */

#include <stdio.h>
#include <p3dgen.h>
#include <drawp3d.h>

/* Macro to ease checking of return codes */
#define ERRCHK( string ) if (!(string)) fprintf(stderr,"ERROR!\n")

/* Light specification information */
static P_Point lightloc= { 0.0, 20.0, 20.0 };
static P_Color lightcolor= { P3D_RGB, 0.7, 0.7, 0.7, 1.0 };
static P_Color ambcolor= { P3D_RGB, 0.3, 0.3, 0.3, 1.0 };

/* Camera specification information */
static P_Point lookfrom= { 0.0, 0.0, 10.0 };
static P_Point lookat= { 0.0, 0.0, 0.0 };
static P_Vector up= { 0.0, 1.0, 0.0 };
static float fovea= 53.0;
static float hither= -5.0;
static float yon= -15.0;

main()
{
  ERRCHK( dp_init_ren("myrenderer","p3d","example_2.p3d",
		      "This string will be ignored") );
  
  ERRCHK( dp_open("mygob") );
  ERRCHK( dp_sphere() );
  ERRCHK( dp_close() );

  ERRCHK( dp_open("mylights") );
  ERRCHK( dp_ambient(&ambcolor) );
  ERRCHK( dp_light(&lightloc, &lightcolor) );
  ERRCHK( dp_close() );

  ERRCHK( dp_camera( "mycamera", &lookfrom, &lookat, &up, fovea,
		    hither, yon ) );

  ERRCHK( dp_snap("mygob","mylights","mycamera") );

  ERRCHK( dp_shutdown() );
}
